### PR TITLE
feat(config): add validation for enum enumerations

### PR DESCRIPTION
**Pull request by Phil91**
## Description

Changed the configuration validation to check IEnumerable<Enum> to only include valid enum values.
Changed configurations to remove case sensitivity

## Why

To remove one possible error when configuring the portal the case sensitivity was removed. All configuration keys are now case insensitive

## Issue

N/A - Jira Ticket: CPLP-2643, CPLP-2698

## Corresponding CD-Pr

[PR-25](https://github.com/eclipse-tractusx/portal-cd/pull/25)

## Checklist

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [X] I have performed [IP checks](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-04#checking-libraries-using-the-eclipse-dash-license-tool) for added or updated 3rd party libraries
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/mailing/Mailing.Template/TemplateSettings.cs [hunk 1]
```diff
@@ -32,29 +32,41 @@ public class TemplateSettings
...
-    public Dictionary<string, TemplateSetting> Templates { get; set; } = null!;
+    public IEnumerable<TemplateInfo> Templates { get; set; } = null!;
 
     public bool Validate()
     {
         var validation = new ConfigurationValidation<TemplateSettings>()
             .NotNull(Templates, () => nameof(Templates));
 
-        return Templates.Values.All(t =>
+        if (Templates.Select(x => x.Name).Distinct().Count() != Templates.Count())
```
Comment (ntruchsess) [inline src/mailing/Mailing.Template/TemplateSettings.cs:42]: =>
```if (Templates.DistinctBy(x => x.Name).Count() != Templates.Count())```
or maybe even better readable:
```if (Templates.GroupBy(x => x.Name).Any(x => x.Count() > 0))```

File: src/marketplace/Services.Service/ServiceSettings.cs [hunk 2]
```diff
@@ -112,11 +121,38 @@ public class ServiceSettings
...
     /// <summary>
     /// Company Admin Roles
     /// </summary>
     [Required]
-    public IDictionary<string, IEnumerable<string>> CompanyAdminRoles { get; set; } = null!;
+    public IEnumerable<UserRoleConfig> CompanyAdminRoles { get; set; } = null!;
+
+    public static bool Validate(ServiceSettings settings)
+    {
+        if (settings.UploadServiceDocumentTypeIds.Select(x => x.DocumentTypeId).Distinct().Count() !=
```
Comment (ntruchsess) [inline src/marketplace/Services.Service/ServiceSettings.cs:134]: see comment in TemplateSettings

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/mailing/Mailing.Template/TemplateSettings.cs (hunk 1 above)
Comment (ntruchsess) [reply src/mailing/Mailing.Template/TemplateSettings.cs:42]: I going to add extension-Methods to the framework to the framework that solves this duplication-detection in a most efficient and generic way. Will have a signature like:
```
IEnumerable<T>.Duplicates(this IEnumerable<T>, IEnumerable<T>)
IEnumerable<T>.DuplicatesBy(this IEnumerable<T>, Func<T,S>, IEnumerable<T>)
```
to be used this way:
```
if (values.DuplicatesBy(x => x.Key).Any()) {...}
```
stay tuned ;-)

Comment (ntruchsess) [reply src/mailing/Mailing.Template/TemplateSettings.cs]: @Phil91 : see https://github.com/eclipse-tractusx/portal-backend/pull/59/commits/1d41118e4d9bf163a12f75ad91bb66ce5c1b8b74

File: src/administration/Administration.Service/BusinessLogic/InvitationBusinessLogic.cs [hunk 3]
```diff
@@ -110,7 +110,7 @@ private async Task ExecuteInvitationInternalAsync(CompanyInvitationData invitati
         IEnumerable<UserRoleData> roleDatas;
         try
         {
-            roleDatas = await _userProvisioningService.GetRoleDatas(_settings.InvitedUserInitialRoles).ToListAsync().ConfigureAwait(false);
+            roleDatas = await _userProvisioningService.GetRoleDatas(_settings.InvitedUserInitialRoles.ToDictionary(x => x.ClientId, x => x.UserRoleNames)).ToListAsync().ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/administration/Administration.Service/BusinessLogic/InvitationBusinessLogic.cs]: changed

File: src/administration/Administration.Service/BusinessLogic/UserBusinessLogic.cs [hunk 4]
```diff
@@ -314,7 +314,7 @@ public async Task<int> AddOwnCompanyUsersBusinessPartnerNumbersAsync(Guid compan
     public async Task<CompanyOwnUserDetails> GetOwnUserDetails(string iamUserId)
     {
         var userRoleIds = await _portalRepositories.GetInstance<IUserRolesRepository>()
-            .GetUserRoleIdsUntrackedAsync(_settings.UserAdminRoles).ToListAsync().ConfigureAwait(false);
+            .GetUserRoleIdsUntrackedAsync(_settings.UserAdminRoles.ToDictionary(x => x.ClientId, x => x.UserRoleNames)).ToListAsync().ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/administration/Administration.Service/BusinessLogic/UserBusinessLogic.cs]: changed

File: src/framework/Framework.Models/Validation/SettingValidation.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    /// <summary>
+    /// Register this options instance for validation of the custom <see cref="EnumEnumerationAttribute"/>.
+    /// </summary>
+    /// <typeparam name="TOptions">The options type to be configured.</typeparam>
+    /// <param name="optionsBuilder">The options builder to add the services to.</param>
+    /// <param name="section">The current configuration section</param>
+    /// <returns>The <see cref="OptionsBuilder{TOptions}"/> so that additional calls can be chained.</returns>
+    public static OptionsBuilder<TOptions> ValidateEnumEnumeration<TOptions>(this OptionsBuilder<TOptions> optionsBuilder, IConfigurationSection section) where TOptions : class
+    {
+        optionsBuilder.Services.AddSingleton<IValidateOptions<TOptions>>(new EnumEnumerableValidation<TOptions>(optionsBuilder.Name, section));
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/framework/Framework.Models/Validation/SettingValidation.cs]: agree, i did change it to transient

File: src/mailing/Mailing.Template/TemplateSettings.cs (hunk 1 above)
Comment (Phil91) [reply src/mailing/Mailing.Template/TemplateSettings.cs]: thanks :) have added the new extension check

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**